fix(completions): add backticks when needed in completions

ckipp01 opened the pull request:
This PR adds in the functionality to add in backticks when needed when
giving back completions and also correctly returning completions when
the user has already started typing a backtick.

Hopefully this will fix the issues in the REPL and also rid the need of Ammonite and Metals to both have custom logic to handle this. Some of the work in here is based off stuff we have in Metals, which itself borrows from Ammonite. There are also parts of @tgodzik's work from https://github.com/lampepfl/dotty/pull/13369 in here and @alexarchambault's work in the early commits of https://github.com/lampepfl/dotty/pull/11794.

Fixes: #4406, #14006

ckipp01 commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -120,16 +135,48 @@ object Completion {
+    !valid
+  end needsBacktick
+
+  private lazy val keywords = Tokens.keywords.map(Tokens.tokenString)
```
I was also a bit unsure about this. Is there any better way to do a token lookup with a string?

prolativ replied:
You could also try `dotty.tools.dotc.core.StdNames.nme.keywords.map(_.toString)`

ckipp01 commented:
Beauty shot showing that it works

![2022-03-01 16 06 04](https://user-images.githubusercontent.com/13974112/156193933-44400071-f606-4942-b648-9860f0bdda85.gif)

ckipp01 commented:
Also note, I was working on https://github.com/lampepfl/dotty/issues/12514 but it introduces other issues. I'm able to detect the `<error>` in the import and that it was caused by an unclosed backtick, but I'm unsure what the compiler should actually return here. In the case given in the actual issue ``import scala.util.chaining.`sc`` there is no need for the back tick here, so while I can get it to return the `scalaUtilChainingOps`, it's not backticked, because it doesn't need to be. In this scenario what's actually desirable? Should we detect that they are trying to backtick something that isn't needed and then backtick that completion item when it's returned? That might work smoother but I do find it a bit odd.

tgodzik commented:
> Also note, I was working on #12514 but it introduces other issues. I'm able to detect the `<error>` in the import and that it was caused by an unclosed backtick, but I'm unsure what the compiler should actually return here. In the case given in the actual issue `` import scala.util.chaining.`sc `` there is no need for the back tick here, so while I can get it to return the `scalaUtilChainingOps`, it's not backticked, because it doesn't need to be. In this scenario what's actually desirable? Should we detect that they are trying to backtick something that isn't needed and then backtick that completion item when it's returned? That might work smoother but I do find it a bit odd mainly because then what do you expect if you try to do a completion with `` .`a `` which returns 10 members, should you backtick all ten of those members? Again that seems odd to me.

I would even remove the backtick if it's no needed. I think it's a rare scenario that someone starts with a backtick if they actually don't need it.

tgodzik reviewed (changes requested):
Some minor comments, but it looks like a good way to go. Thanks!

prolativ commented:
I checked out to this branch and played with it a bit in REPL. Having this code defined
```scala
object Foo:
  val foo = 1
  val `foo-bar` = 2
  val `bar` = 3
  val `match` = 4
  val `has space` = 5

val `foo-baz` = 6
```
I tried completions with prefixes:
`Foo.fo` -> suggests `foo` but not `` `foo-bar` ``
``Foo.`fo`` -> suggests `` `foo-bar` `` but not `` `foo` ``
``Foo.`ba`` -> doesn't suggest anything
``Foo.`has `` (with a trailing space) -> doesn't suggest anything
`` `fo`` -> doesn't suggest anything

prolativ commented:
Even if we include backticks in `label`s of `Completion`s I think it would be better to strip them when displaying the list of suggestions in REPL (similarly to how this is displayed in Metals)

ckipp01 commented on compiler/src/dotty/tools/repl/ReplDriver.scala:
```diff
@@ -35,6 +35,8 @@ import org.jline.reader._
 import scala.annotation.tailrec
 import scala.collection.JavaConverters._
 import scala.util.Using
+import dotty.tools.dotc.util.Chars
```
Ah yes, `Scanner` isn't used here but Char and Tokens are.

tgodzik replied:
How so? There seem to be changes to this file aside from the imports :thinking:

ckipp01 replied:
Ahhh sorry, I was looking in the wrong file! I'll fix this.

ckipp01 commented on language-server/test/dotty/tools/languageserver/CompletionTest.scala:
```diff
@@ -1023,4 +1023,59 @@ class CompletionTest {
+           |    case `back-tick`
+           |    case `match`
+           |  
+           |  val x = Bar.`back${m1}"""
```
So this one is tricky. Mainly because when we trigger a completion here and we look at the path that is returned for `Interactive.pathTo`, it's actually empty because it no longer thinks the position you are looking at is part of anything. I have no idea how to get around that.

tgodzik replied:
I would say it's a really niche case and I would just ignore it. It's highly unlikely that someone will want a completion on a space

ckipp01 commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -88,13 +91,22 @@ object Completion {
           completionPrefix(selector :: Nil, pos)
         }.getOrElse("")
 
+      // We special case Select here because we want to determine if the name
```
Yea I'm actually working with figuring this out for `Ident` as well. However one difference I noticed that I don't fully get is that the `name` of a `Select` is a nospan when it's an `<error>`, however the name of an `Ident` when it's an `<error>` has a span. This makes getting the prefix different for each since here in the Select it's a bit hacky. How come one `<error>` is a no span and the other isn't?

ckipp01 commented:
> I checked out to this branch and played with it a bit in REPL. Having this code defined
> 
> ```scala
> object Foo:
>   val foo = 1
>   val `foo-bar` = 2
>   val `bar` = 3
>   val `match` = 4
>   val `has space` = 5
> 
> val `foo-baz` = 6
> ```
> 
> I tried completions with prefixes: `Foo.fo` -> suggests `foo` but not `` `foo-bar` `` `` Foo.`fo `` -> suggests `` `foo-bar` `` but not `` `foo` `` `` Foo.`ba `` -> doesn't suggest anything `` Foo.`has `` (with a trailing space) -> doesn't suggest anything `` `fo `` -> doesn't suggest anything

So I've addressed most of these but still have one or two more to go.

> Foo.fo -> suggests foo but not `foo-bar`

This is actually the REPL trying to be smart and recognizing that what you have doesn't start with a `` ` `` so it's not going to offer you that. However I added a test showing that it is actually returned.

> Foo.`fo -> suggests `foo-bar` but not `foo`

This one is now fixed, and really now anytime there is a backtick even when not needed it will still complete it and add the backticks.

> Foo.`ba -> doesn't suggest anything

This one is also now fixed

> Foo.`has (with a trailing space) -> doesn't suggest anything

This one I addressed above in a comment as it's a whole other beast that might have to be fixed separately since the issue is totally different with `pathTo` not returning anything.

> `fo -> doesn't suggest anything

This one isn't yet fixed but I'll add another commit to address it.

ckipp01 commented:
> Even if we include backticks in labels of Completions I think it would be better to strip them when displaying the list of suggestions in REPL (similarly to how this is displayed in Metals)

Done

som-snytt commented:
The scala 2 ticket is from 2013. https://github.com/scala/bug/issues/6919

I'm unable to add heart emoji reaction more than once.

prolativ commented:
The problem with ``Foo.`has ``(with a trailing space) indeed seems to be deeper. It looks like the parser strips whitespaces from the end of the input so the position of the cursor during completion is outside of the parsed tree. Similarly there are no completions for `Foo. `(with a trailing space).

Could you elaborate on the problem in REPL for `Foo.fo`? It seems to work correctly in case of `Foo.ma`, where is suggests ``Foo.`match` ``. However if you add `val maatch = 7` as a member of `Foo` then only `maatch` is suggested and `match` isn't

ckipp01 commented:
> Could you elaborate on the problem in REPL for `Foo.fo`? It seems to work correctly in case of `Foo.ma`, where is suggests `` Foo.`match` ``. However if you add `val maatch = 7` as a member of `Foo` then only `maatch` is suggested and `match` isn't

Sure, you can see it illustrated in the `backticksCompleteBoth` both. So in the case of `Foo.fo` both `foo` and `` `foo-bar` `` _are_ being returned, but since jline sees that you aren't starting with a `` ` `` it just filters it out and only offers foo. The same things happens with the match example. It actually does return both `maatch` and `` `match` `` but since you aren't starting with a `` ` `` jline filters the `` `match` `` out.  This seems to differ when there is only one completion option returned and why trying to complete `Foo.ma`  will complete `` `match` `` when it's the only completion item returned. All that to say, this seems to be something that jline is doing, not something that the completion logic that these changes are doing. I can try to dig into that further in the future, but trying to limit what I all touch in this.

tgodzik commented:
> What's returned here from `pathTo` is:

I looks like the enclosing class is only returned, I think it's a case that no actual Ident is being created and it's something we might want to fix in the parser. We should return an Ident with an error name in that case. We could do it in a follow up though.

tgodzik reviewed (approved):
Looks awesome! I would say that any improvement in case of backticks is a good way forward, but I would not focus on this as it is a super niche use case with some of the edge cases.

ckipp01 commented:
> Looks awesome! I would say that any improvement in case of backticks is a good way forward, but I would not focus on this as it is a super niche use case with some of the edge cases.

Cool, glad to hear this 😄 . I spent a big chunk of the day today looking at the edge cases here and worked a bit with @bishabosha on it as well. The issue seems to be that when we encounter unclosed backticks in the parser depending on where they are, we create various different things. So there are two more cases locally I actually have a fix for, but it's more of hack than a fix.  To give a couple some examples:

```scala

val `foo-bar` = 3

val x = `foo@@
```

`` `foo`` here ends up being a `Literal(Constant(null))`. So I can capture that in the various places, but it's messy. Furthermore if this is a `Type` instead of a `Term` like:

```scala
type Foo = `Lis@@
```

`` `Lis`` here will end up being a `TypeDef(Foo,Ident(<error>))`. We can capture that as well, but you start to see that there are various different trees we need to special case all over the place to ensure that baktick completions work.

One thing we tried was to actually tag this with a `StickyKey` attachment in the parser so that we could just check later on, but for some reason the attachment was no longer there when I was checking for it.

All that to say, I think maybe this should be fixed in the parser and try to unify the way we handle unclosed backticks in a way that can easily be determined later on in a similar way. The fix that is here currently _should_ take care of probably the majority of cases users will actually hit on, but these niche ones are difficult. Ha, I've been hitting more walls than expected with this 😅

tgodzik commented:
@prolativ what do you say that we merge this as it's already a great improvement? We should but the remaining edge cases into an issue.

bishabosha commented:
We had a small discussion and it could be worth trying to make a new Token for an unfinished backticked identifier, which we can then make a valid tree for

ckipp01 commented:
> @prolativ what do you say that we merge this as it's already a great improvement? We should but the remaining edge cases into an issue.

I'm all for this, and yea I can extract the remaining edgecases out into issues with details since I think the way forward is what @bishabosha recommends, but I'd rather not mix that with this pr.

prolativ commented:
Ok. I started to dig a bit into the corner case with 

```scala
object Foo:
  val foo = 1
  val `foo-bar` = 2
  
Foo.`fo<TAB>
```
myself but indeed this seems very tricky so let's get this merged first
